README.md
# Random Keygen

Enhance security with RandomKeygen: Easily generate strong passwords for your applications, services, and devices.

CHANGELOG.md
# Random Keygen Changelog

## [Initial Version] - 2023-09-17

src/utils/key-gen.ts
import { randomBytes } from "crypto";

const lowerCase = "abcdefghijklmnopqrstuvwxyz";
const upperCase = "ABCDEFGHIJKLMNOPQRSTUVWXYZ";
const numbers = "1234567890";
const special = "`~!@#$%^&*()-=_+[]{}|;':\",./<>?";
const hex = "123456789ABCDEF";

interface Options {
  length: number;
  useLowerCase?: boolean;
  useUpperCase?: boolean;
  useNumbers?: boolean;
  useSpecial?: boolean;
  useHex?: boolean;
}

function keyGen(ops: Options): string {
  let chars = "";
  let key = "";

  if (ops.useLowerCase) chars += lowerCase;
  if (ops.useUpperCase) chars += upperCase;
  if (ops.useNumbers) chars += numbers;
  if (ops.useSpecial) chars += special;
  if (ops.useHex) chars += hex;

  for (let i = 0; i < ops.length; i++) {
    key += chars[Math.floor((randomBytes(1).readUInt8(0) / 255) * chars.length)];
  }

  return key;
}

export enum KeyEnum {
  "decent",
  "strong",
  "ft_knox",
  "ci_key",
  "wpa_160",
  "wpa_504",
  "wep64",
  "wep128",
  "wep152",
  "wep256",
}

function getOptionsForKey(strength: KeyEnum): Options {
  switch (strength) {
    case KeyEnum.decent:
      return { length: 10 };
    case KeyEnum.strong:
      return { length: 15, useSpecial: true };
    case KeyEnum.ft_knox:
      return { length: 30, useSpecial: true };
    case KeyEnum.ci_key:
      return { length: 32 };
    case KeyEnum.wpa_160:
      return { length: 20, useSpecial: true };
    case KeyEnum.wpa_504:
      return { length: 63, useSpecial: true };
    case KeyEnum.wep64:
      return {
        length: 5,
        useNumbers: false,
        useLowerCase: false,
        useUpperCase: false,
        useSpecial: false,
        useHex: true,
      };
    case KeyEnum.wep128:
      return {
        length: 13,
        useNumbers: false,
        useLowerCase: false,
        useUpperCase: false,
        useSpecial: false,
        useHex: true,
      };
    case KeyEnum.wep152:
      return {
        length: 16,
        useNumbers: false,
        useLowerCase: false,
        useUpperCase: false,
        useSpecial: false,
        useHex: true,
      };
    case KeyEnum.wep256:
      return {
        length: 29,
        useNumbers: false,
        useLowerCase: false,
        useUpperCase: false,
        useSpecial: false,
        useHex: true,
      };
    default:
      throw Error(`Invalid key type`);
  }
}

export function getKey(strength: KeyEnum) {
  return keyGen(getOptionsForKey(strength));
}


src/utils/constants.ts
import { KeyEnum } from "./generator";

export type KeySection = {
  title: string;
  description: string;
  type: KeyEnum;
};

export const keySections: KeySection[] = [
  {
    title: "Memorable Passwords",
    description: "Decently secure passwords. Balanced for memorability and security. Suitable for daily use.",
    type: KeyEnum.decent,
  },
  {
    title: "Strong Passwords",
    description: "Robust security with a mix of different character types. Ideal for protecting sensitive data.",
    type: KeyEnum.strong,
  },
  {
    title: "Fort Knox Passwords",
    description: "Ultimate level of security. Complex and lengthy. Ideal for protecting highly confidential data.",
    type: KeyEnum.ft_knox,
  },
  {
    title: "CodeIgniter Encryption Keys",
    description: "Solid 32-char keys for use with CodeIgniter PHP framework's encryption services.",
    type: KeyEnum.ci_key,
  },
  {
    title: "160-bit WPA Key",
    description: "Wi-Fi Protected Access (WPA) keys with higher security. Intended for securing wireless networks.",
    type: KeyEnum.wpa_160,
  },
  {
    title: "504-bit WPA Key",
    description: "Maximum length key for WPA. Ensuring advanced level security for Wi-Fi networks.",
    type: KeyEnum.wpa_504,
  },
  {
    title: "64-bit WEP Keys",
    description:
      "Entry-level Wireless Encryption Protocol (WEP) keys. Best suited for older devices or minimal security needs.",
    type: KeyEnum.wep64,
  },
  {
    title: "128-bit WEP Keys",
    description: "Advanced level WEP keys providing stronger security than the 64-bit version.",
    type: KeyEnum.wep128,
  },
  {
    title: "152-bit WEP Keys",
    description: "Specialized WEP keys providing above average security. Compatible with specific devices only.",
    type: KeyEnum.wep152,
  },
  {
    title: "256-bit WEP Keys",
    description: "The strongest of the WEP keys for maximum wireless security, but compatible with very few devices.",
    type: KeyEnum.wep256,
  },
];


src/utils/generator.ts
import { randomNumber } from "./random";

const lowercase = "abcdefghijklmnopqrstuvwxyz";
const uppercase = "ABCDEFGHIJKLMNOPQRSTUVWXYZ";
const numbers = "0123456789";
const symbols = '!@#$%^&*()+_-=}{[]|:;"/?.><,`~';
const hex = "123456789ABCDEF";

const generate = (opts: Options, pool: string): string => {
  let password = "";
  const optionsLength = opts.length;
  const poolLength = pool.length;

  for (let i = 0; i < optionsLength; i++) {
    password += pool[randomNumber(poolLength)];
  }

  return password;
};

type Options = {
  length: number;
  useNumbers?: boolean;
  useSymbols?: boolean;
  useUpperCase?: boolean;
  useLowerCase?: boolean;
  useHex?: boolean;
};

const generateKey = (
  opts: Options = {
    length: 10,
    useNumbers: false,
    useLowerCase: true,
    useUpperCase: true,
    useSymbols: false,
    useHex: false,
  }
): string => {
  let pool = "";

  if (opts.useNumbers) pool += numbers;
  if (opts.useLowerCase) pool += lowercase;
  if (opts.useUpperCase) pool += uppercase;
  if (opts.useSymbols) pool += symbols;
  if (opts.useHex) pool += hex;

  if (!pool) throw new Error("At least one rule for pools must be true");

  return generate(opts, pool);
};

export enum KeyEnum {
  "decent",
  "strong",
  "ft_knox",
  "ci_key",
  "wpa_160",
  "wpa_504",
  "wep64",
  "wep128",
  "wep152",
  "wep256",
}

function getOptionsForKey(strength: KeyEnum): Options {
  switch (strength) {
    case KeyEnum.decent:
      return {
        length: 10,
        useNumbers: true,
        useLowerCase: true,
        useUpperCase: true,
      };
    case KeyEnum.strong:
      return {
        length: 15,
        useLowerCase: true,
        useUpperCase: true,
        useSymbols: true,
      };
    case KeyEnum.ft_knox:
      return {
        length: 30,
        useLowerCase: true,
        useUpperCase: true,
        useSymbols: true,
      };
    case KeyEnum.ci_key:
      return {
        length: 32,
        useNumbers: true,
        useLowerCase: true,
        useUpperCase: true,
      };
    case KeyEnum.wpa_160:
      return {
        length: 20,
        useLowerCase: true,
        useUpperCase: true,
        useSymbols: true,
      };
    case KeyEnum.wpa_504:
      return {
        length: 63,
        useSymbols: true,
      };
    case KeyEnum.wep64:
      return {
        length: 5,
        useHex: true,
      };
    case KeyEnum.wep128:
      return {
        length: 13,
        useHex: true,
      };
    case KeyEnum.wep152:
      return {
        length: 16,
        useHex: true,
      };
    case KeyEnum.wep256:
      return {
        length: 29,
        useHex: true,
      };
    default:
      throw Error(`Invalid key type`);
  }
}

export function getKey(strength: KeyEnum) {
  return generateKey(getOptionsForKey(strength));
}
